packages/engine/src/input/systems/ClientInputSystem.ts

```diff
@@ -1,177 +1,212 @@
+  addListener(canvas, 'contextmenu', preventDefault)
 
-  const prevData = new Map<InputAlias, InputValue>()
-  for (const [key, value] of inputComponent.data) prevData.set(key, value)
+  let lastPrimaryClickDown = 0
+  let lastAuxiliaryClickDown = 0
+  let lastSecondaryClickDown = 0
 
-  inputComponent.data.clear()
+  const keyState = getState(ButtonInputState)
 
-  // apply the input mappings
-  for (const [input, mapping] of inputComponent.schema.inputMap) {
-    if (typeof input === 'object') {
-      processCombinationLifecycle(inputComponent, prevData, mapping, input)
-    } else {
-      if (Engine.instance.currentWorld.inputState.has(input)) {
-        inputComponent.data.set(mapping, JSON.parse(JSON.stringify(Engine.instance.currentWorld.inputState.get(input))))
+  const handleMouseClick = (event: MouseEvent) => {
+    const down = event.type === 'mousedown' || event.type === 'touchstart'
+
+    let button: ButtonTypes = 'PrimaryClick'
+    if (event.button === 1) button = 'AuxiliaryClick'
+    else if (event.button === 2) button = 'SecondaryClick'
+
+    if (down) keyState[button].set(true)
+    else keyState[button].set(none)
+
+    if (down) {
+      const now = Date.now()
+      if (button === 'PrimaryClick') {
+        if (now - lastPrimaryClickDown < 200 && now - lastPrimaryClickDown > 50)
+          keyState['PrimaryDoubleClick'].set(true)
+        lastPrimaryClickDown = now
       }
+
+      if (button === 'AuxiliaryClick') {
+        if (now - lastAuxiliaryClickDown < 200 && now - lastAuxiliaryClickDown > 50)
+          keyState['AuxiliaryDoubleClick'].set(true)
+        lastAuxiliaryClickDown = now
+      }
+
+      if (button === 'SecondaryClick') {
+        if (now - lastSecondaryClickDown < 200 && now - lastSecondaryClickDown > 50)
+          keyState['SecondaryDoubleClick'].set(true)
+        lastSecondaryClickDown = now
+      }
+    } else {
+      if (button === 'PrimaryClick' && keyState['PrimaryDoubleClick'].value) keyState['PrimaryDoubleClick'].set(none)
+      if (button === 'PrimaryClick' && keyState['PrimaryMove'].value) keyState['PrimaryMove'].set(none)
+
+      if (button === 'SecondaryClick' && keyState['SecondaryDoubleClick'].value)
+        keyState['SecondaryDoubleClick'].set(none)
+      if (button === 'SecondaryClick' && keyState['SecondaryMove'].value) keyState['SecondaryMove'].set(none)
+
+      if (button === 'AuxiliaryClick' && keyState['AuxiliaryDoubleClick'].value)
+        keyState['AuxiliaryDoubleClick'].set(none)
+      if (button === 'AuxiliaryClick' && keyState['AuxiliaryMove'].value) keyState['AuxiliaryMove'].set(none)
     }
   }
 
-  // now that we have the data mapped, run the behaviors
-  for (const [key, value] of inputComponent.data) {
-    if (inputComponent.schema.behaviorMap.has(key)) {
-      inputComponent.schema.behaviorMap.get(key)!(entity, key, value)
+  const handleMouseMove = (event: MouseEvent) => {
+    if (keyState['PrimaryClick'].value) keyState['PrimaryMove'].set(true)
```

Comment:
why are we resetting these values to their current state?

Reply:
Read again, this is primary click and primary move